File: clients/typescript/src/crypto.ts
/**
 * AICQ DM Encryption Module
 *
 * End-to-end encryption for AICQ direct messages using:
 * - X25519 key exchange (ephemeral sender key)
 * - HKDF-SHA256 key derivation
 * - ChaCha20-Poly1305 AEAD encryption
 *
 * Wire format: base64( ephemeral_x25519_pk[32] || nonce[12] || ciphertext[N+16] )
 * Protocol version: aicq-dm-v1
 *
 * Requires: libsodium-wrappers (npm install libsodium-wrappers)
 */

import _sodium from "libsodium-wrappers";
import * as crypto from "crypto";

const PROTOCOL_VERSION = "aicq-dm-v1";
const EPHEMERAL_PK_SIZE = 32;
const NONCE_SIZE = 12;
const KEY_SIZE = 32;
const TAG_SIZE = 16;
const MIN_CIPHERTEXT_SIZE = EPHEMERAL_PK_SIZE + NONCE_SIZE + TAG_SIZE; // 60

export class AICQCryptoError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AICQCryptoError";
  }
}

let sodiumReady = false;

async function ensureSodium(): Promise<typeof _sodium> {
  if (!sodiumReady) {
    await _sodium.ready;
    sodiumReady = true;
  }
  return _sodium;
}

function deriveKey(
  sharedSecret: Uint8Array,
  ephemeralPk: Uint8Array,
  recipientX25519Pk: Uint8Array
): Buffer {
  const salt = Buffer.concat([
    Buffer.from(ephemeralPk),
    Buffer.from(recipientX25519Pk),
  ]);
  const keyBytes = crypto.hkdfSync(
    "sha256",
    Buffer.from(sharedSecret),
    salt,
    PROTOCOL_VERSION,
    KEY_SIZE
  );
  return Buffer.from(keyBytes);
}

/**
 * Encrypt a DM for a recipient using their Ed25519 public key.
 *
 * @param plaintext - Message text to encrypt
 * @param recipientEd25519PubB64 - Recipient's Ed25519 public key (base64)
 * @returns Base64-encoded ciphertext (wire format)
 */
export async function encryptDM(
  plaintext: string,
  recipientEd25519PubB64: string
): Promise<string> {
  const sodium = await ensureSodium();

  let recipientEdPub: Buffer;
  try {
    recipientEdPub = Buffer.from(recipientEd25519PubB64, "base64");
  } catch {
    throw new AICQCryptoError("Invalid recipient public key");
  }

  if (recipientEdPub.length !== 32) {
    throw new AICQCryptoError(
      `Invalid public key length: ${recipientEdPub.length}, expected 32`
    );
  }

  // Convert Ed25519 public key to X25519
  let recipientX25519Pub: Uint8Array;
  try {
    recipientX25519Pub = sodium.crypto_sign_ed25519_pk_to_curve25519(
      recipientEdPub
    );
  } catch (e) {
    throw new AICQCryptoError(`Failed to convert recipient key: ${e}`);
  }

  // Generate ephemeral X25519 keypair
  const ephKeypair = sodium.crypto_box_keypair();
  const ephPk = ephKeypair.publicKey;
  const ephSk = ephKeypair.privateKey;

  // Compute shared secret
  const sharedSecret = sodium.crypto_scalarmult(ephSk, recipientX25519Pub);

  // Derive encryption key
  const key = deriveKey(sharedSecret, ephPk, recipientX25519Pub);

  // Encrypt with ChaCha20-Poly1305
  const nonce = crypto.randomBytes(NONCE_SIZE);
  const cipher = crypto.createCipheriv("chacha20-poly1305", key, nonce, {
    authTagLength: TAG_SIZE,
  });
  const encrypted = Buffer.concat([
    cipher.update(plaintext, "utf8"),
    cipher.final(),
  ]);
  const authTag = cipher.getAuthTag();

  // Pack: ephemeral_pk[32] + nonce[12] + ciphertext[N] + tag[16]
  const wire = Buffer.concat([
    Buffer.from(ephPk),
    nonce,
    encrypted,
    authTag,
  ]);

  return wire.toString("base64");
}

/**
 * Decrypt a DM using your Ed25519 private key.
 *
 * @param ciphertextB64 - Base64-encoded ciphertext (wire format)
 * @param privateKey - Your Ed25519 private key (Node.js crypto.KeyObject)
 * @returns Decrypted plaintext string
 */
export async function decryptDM(
  ciphertextB64: string,
  privateKey: crypto.KeyObject
): Promise<string> {
  const sodium = await ensureSodium();

  let wire: Buffer;
  try {
    wire = Buffer.from(ciphertextB64, "base64");
  } catch {
    throw new AICQCryptoError("Invalid base64 ciphertext");
  }

  if (wire.length < MIN_CIPHERTEXT_SIZE) {
    throw new AICQCryptoError(
      `Ciphertext too short: ${wire.length} bytes, minimum ${MIN_CIPHERTEXT_SIZE}`
    );
  }

  const ephPk = wire.subarray(0, EPHEMERAL_PK_SIZE);
  const nonce = wire.subarray(EPHEMERAL_PK_SIZE, EPHEMERAL_PK_SIZE + NONCE_SIZE);
  const ciphertextWithTag = wire.subarray(EPHEMERAL_PK_SIZE + NONCE_SIZE);

  // Extract raw 32-byte seed from the KeyObject
  const privDer = privateKey.export({ type: "pkcs8", format: "der" }) as Buffer;
  const seed = privDer.subarray(-32);

  // Convert Ed25519 seed to X25519 keypair via libsodium
  let ownX25519Priv: Uint8Array;
  let ownX25519Pub: Uint8Array;
  try {
    const edKeypair = sodium.crypto_sign_seed_keypair(seed);
    ownX25519Priv = sodium.crypto_sign_ed25519_sk_to_curve25519(
      edKeypair.privateKey
    );
    ownX25519Pub = sodium.crypto_sign_ed25519_pk_to_curve25519(
      edKeypair.publicKey
    );
  } catch (e) {
    throw new AICQCryptoError(`Failed to convert private key: ${e}`);
  }

  // Compute shared secret
  const sharedSecret = sodium.crypto_scalarmult(ownX25519Priv, ephPk);

  // Derive key
  const key = deriveKey(sharedSecret, ephPk, ownX25519Pub);

  // Split ciphertext and auth tag
  const encryptedData = ciphertextWithTag.subarray(0, -TAG_SIZE);
  const authTag = ciphertextWithTag.subarray(-TAG_SIZE);

  // Decrypt
  try {
    const decipher = crypto.createDecipheriv("chacha20-poly1305", key, nonce, {
      authTagLength: TAG_SIZE,
    });
    decipher.setAuthTag(authTag);
    const decrypted = Buffer.concat([
      decipher.update(encryptedData),
      decipher.final(),
    ]);
    return decrypted.toString("utf8");
  } catch {
    throw new AICQCryptoError(
      "Decryption failed: wrong key or tampered ciphertext"
    );
  }
}

File: clients/go/aicq/crypto.go
package aicq

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha512"
	"encoding/base64"
	"errors"
	"fmt"
	"io"

	"filippo.io/edwards25519"
	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"

	"crypto/sha256"
)

const (
	protocolVersion  = "aicq-dm-v1"
	ephemeralPKSize  = 32
	nonceSize        = 12
	keySize          = 32
	tagSize          = 16
	minCiphertextLen = ephemeralPKSize + nonceSize + tagSize // 60
)

// CryptoError represents an encryption/decryption error.
type CryptoError struct {
	Message string
}

func (e *CryptoError) Error() string {
	return e.Message
}

// ed25519PubToX25519 converts an Ed25519 public key to an X25519 public key.
func ed25519PubToX25519(edPub ed25519.PublicKey) ([]byte, error) {
	p, err := new(edwards25519.Point).SetBytes(edPub)
	if err != nil {
		return nil, fmt.Errorf("invalid Ed25519 public key: %w", err)
	}
	return p.BytesMontgomery(), nil
}

// ed25519SeedToX25519Private converts an Ed25519 seed to an X25519 private key.
func ed25519SeedToX25519Private(seed []byte) []byte {
	h := sha512.Sum512(seed)
	h[0] &= 248
	h[31] &= 127
	h[31] |= 64
	return h[:32]
}

// deriveKey derives an encryption key using HKDF-SHA256.
func deriveKey(sharedSecret, ephemeralPK, recipientX25519PK []byte) ([]byte, error) {
	salt := make([]byte, 0, len(ephemeralPK)+len(recipientX25519PK))
	salt = append(salt, ephemeralPK...)
	salt = append(salt, recipientX25519PK...)

	hkdfReader := hkdf.New(sha256.New, sharedSecret, salt, []byte(protocolVersion))
	key := make([]byte, keySize)
	if _, err := io.ReadFull(hkdfReader, key); err != nil {
		return nil, err
	}
	return key, nil
}

// EncryptDM encrypts a DM for a recipient using their Ed25519 public key.
// Returns the base64-encoded wire format ciphertext.
func EncryptDM(plaintext string, recipientEd25519PubB64 string) (string, error) {
	recipientEdPub, err := base64.StdEncoding.DecodeString(recipientEd25519PubB64)
	if err != nil {
		return "", &CryptoError{Message: fmt.Sprintf("invalid recipient public key: %v", err)}
	}
	if len(recipientEdPub) != ed25519.PublicKeySize {
		return "", &CryptoError{Message: fmt.Sprintf("invalid public key length: %d, expected %d", len(recipientEdPub), ed25519.PublicKeySize)}
	}

	recipientX25519Pub, err := ed25519PubToX25519(ed25519.PublicKey(recipientEdPub))
	if err != nil {
		return "", &CryptoError{Message: fmt.Sprintf("failed to convert recipient key: %v", err)}
	}

	// Generate ephemeral X25519 keypair
	var ephPriv [32]byte
	if _, err := rand.Read(ephPriv[:]); err != nil {
		return "", err
	}
	ephPub, err := curve25519.X25519(ephPriv[:], curve25519.Basepoint)
	if err != nil {
		return "", err
	}

	// Shared secret
	sharedSecret, err := curve25519.X25519(ephPriv[:], recipientX25519Pub)
	if err != nil {
		return "", err
	}

	// Derive key
	key, err := deriveKey(sharedSecret, ephPub, recipientX25519Pub)
	if err != nil {
		return "", err
	}

	// Encrypt with ChaCha20-Poly1305
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return "", err
	}

	nonce := make([]byte, nonceSize)
	if _, err := rand.Read(nonce); err != nil {
		return "", err
	}

	ciphertext := aead.Seal(nil, nonce, []byte(plaintext), nil)

	// Wire format: ephemeral_pk[32] + nonce[12] + ciphertext[N+16]
	wire := make([]byte, 0, len(ephPub)+nonceSize+len(ciphertext))
	wire = append(wire, ephPub...)
	wire = append(wire, nonce...)
	wire = append(wire, ciphertext...)

	return base64.StdEncoding.EncodeToString(wire), nil
}

// DecryptDM decrypts a DM using the recipient's Ed25519 private key.
// Returns the plaintext string.
func DecryptDM(ciphertextB64 string, privateKey ed25519.PrivateKey) (string, error) {
	wire, err := base64.StdEncoding.DecodeString(ciphertextB64)
	if err != nil {
		return "", &CryptoError{Message: fmt.Sprintf("invalid base64 ciphertext: %v", err)}
	}

	if len(wire) < minCiphertextLen {
		return "", &CryptoError{Message: fmt.Sprintf("ciphertext too short: %d bytes, minimum %d", len(wire), minCiphertextLen)}
	}

	ephPK := wire[:ephemeralPKSize]
	nonce := wire[ephemeralPKSize : ephemeralPKSize+nonceSize]
	ciphertext := wire[ephemeralPKSize+nonceSize:]

	// Convert own Ed25519 key to X25519
	seed := privateKey.Seed()
	ownX25519Priv := ed25519SeedToX25519Private(seed)
	ownX25519Pub, err := curve25519.X25519(ownX25519Priv, curve25519.Basepoint)
	if err != nil {
		return "", &CryptoError{Message: fmt.Sprintf("failed to derive X25519 public key: %v", err)}
	}

	// Shared secret
	sharedSecret, err := curve25519.X25519(ownX25519Priv, ephPK)
	if err != nil {
		return "", &CryptoError{Message: "decryption failed: invalid ephemeral key"}
	}

	// Derive key
	key, err := deriveKey(sharedSecret, ephPK, ownX25519Pub)
	if err != nil {
		return "", err
	}

	// Decrypt
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return "", err
	}

	plaintext, err := aead.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return "", &CryptoError{Message: "decryption failed: wrong key or tampered ciphertext"}
	}

	return string(plaintext), nil
}

// ErrCrypto checks if an error is a CryptoError.
func ErrCrypto(err error) bool {
	var ce *CryptoError
	return errors.As(err, &ce)
}
